# Predicción de ocurrencia de infarto cerebral 

## 1. Análisis exploratorio de los datos 

In [2]:
# importación de librerías
import pandas as pd
import numpy as np

In [3]:
# lectura del csv
df = pd.read_csv ("stroke.csv")

In [4]:
# muestra de las 5 primeras filas
df.head()

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


In [5]:
# dimensiones del dataframe 
df.shape

(5110, 12)

El dataframe tiene 12 columnas con las variables de 5110 individuos. 

In [6]:
#información sobre las variables
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   str    
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   str    
 6   work_type          5110 non-null   str    
 7   Residence_type     5110 non-null   str    
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   str    
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), str(5)
memory usage: 479.2 KB


### 1.1 Las variables

### id

In [ ]:
# Devuelve True si hay al menos un ID duplicado, False si todos son únicos
df["id"].duplicated().any()

np.False_

#### gender

In [ ]:
# Muestra un array con las respuestas únicas
df["gender"].unique()

<StringArray>
['Male', 'Female', 'Other']
Length: 3, dtype: str

In [ ]:
# Frecuencia absoluta de cada respuesta
df["gender"].value_counts()

gender
Female    2994
Male      2115
Other        1
Name: count, dtype: int64

In [ ]:
# Frecuencia relativa (porcentajes)
df["gender"].value_counts(normalize=True) * 100

gender
Female    58.590998
Male      41.389432
Other      0.019569
Name: proportion, dtype: float64

### age

In [ ]:
# Redondea al entero más cercano y convierte el tipo de dato a entero (int64)
df["age"] = df["age"].round().astype(int)

In [ ]:
# Devuelve una lista/array con los valores únicos de edad ordenados
np.sort(df["age"].unique())

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82])

### hypertension

In [ ]:
# valores únicos
df["hypertension"].unique()

array([0, 1])

In [24]:
# Frecuencia relativa (porcentajes)
df["hypertension"].value_counts(normalize=True) * 100

hypertension
0    90.254403
1     9.745597
Name: proportion, dtype: float64

### heart_disease

In [ ]:
# valores único
df["heart_disease"].unique()

array([1, 0])

In [26]:
# Frecuencia relativa (porcentajes)
df["heart_disease"].value_counts(normalize=True) * 100

heart_disease
0    94.598826
1     5.401174
Name: proportion, dtype: float64

### ever_married

In [27]:
df["ever_married"].unique()

<StringArray>
['Yes', 'No']
Length: 2, dtype: str

In [28]:
# Mapeo directo de Yes/No a 1/0
df["ever_married"] = df["ever_married"].map({"Yes": 1, "No": 0})

In [30]:
# Frecuencia relativa (porcentajes)
df["ever_married"].value_counts(normalize=True) * 100

ever_married
1    65.616438
0    34.383562
Name: proportion, dtype: float64

### work_type

In [31]:
df["work_type"].unique()

<StringArray>
['Private', 'Self-employed', 'Govt_job', 'children', 'Never_worked']
Length: 5, dtype: str

In [32]:
df["work_type"].value_counts(normalize = True) * 100

work_type
Private          57.240705
Self-employed    16.027397
children         13.444227
Govt_job         12.857143
Never_worked      0.430528
Name: proportion, dtype: float64

In [37]:
# Ver resumen estadístico de la edad para la categoría 'children'
df[df["work_type"] == "children"]["age"].describe()

count    687.000000
mean       6.841339
std        4.536491
min        0.000000
25%        2.000000
50%        6.000000
75%       11.000000
max       16.000000
Name: age, dtype: float64

In [ ]:
# Frecuencia absoluta y relativa de infarto cerebral 
pd.crosstab(df["work_type"], df["stroke"], margins=True, normalize="index") * 100

stroke,0,1
work_type,,
Govt_job,94.977169,5.022831
Never_worked,100.000000,0.000000
Private,94.905983,5.094017
Self-employed,92.063492,7.936508
children,99.708879,0.291121
All,95.127202,4.872798


### Residence-type

In [33]:
df["Residence_type"].unique()

<StringArray>
['Urban', 'Rural']
Length: 2, dtype: str

In [38]:
# Ponemos el nombre de la columna en minuscula para normalizar 
df = df.rename(columns={"Residence_type": "residence_type"})

In [39]:
# frecuencias relativas
df["residence_type"].value_counts(normalize = True) * 100

residence_type
Urban    50.802348
Rural    49.197652
Name: proportion, dtype: float64

### avg_glucose_level

In [40]:
df["avg_glucose_level"].describe()

count    5110.000000
mean      106.147677
std        45.283560
min        55.120000
25%        77.245000
50%        91.885000
75%       114.090000
max       271.740000
Name: avg_glucose_level, dtype: float64

### bmi

In [41]:
df ["bmi"].describe()

count    4909.000000
mean       28.893237
std         7.854067
min        10.300000
25%        23.500000
50%        28.100000
75%        33.100000
max        97.600000
Name: bmi, dtype: float64

In [ ]:
# Presencia de valores nulos
print(f"Valores nulos en BMI: {df["bmi"].isnull().sum()} ({df["bmi"].isnull().mean()*100:.2f}%)")

Valores nulos en BMI: 201 (3.93%)


### smoking_status

In [45]:
df["smoking_status"].unique()

<StringArray>
['formerly smoked', 'never smoked', 'smokes', 'Unknown']
Length: 4, dtype: str

In [46]:
# frecuencias relativas en porcentaje
df["smoking_status"].value_counts(normalize=True) * 100

smoking_status
never smoked       37.025440
Unknown            30.215264
formerly smoked    17.318982
smokes             15.440313
Name: proportion, dtype: float64

### stroke

In [47]:
df["stroke"].unique()

array([1, 0])

In [48]:
# frecuencias relativas en porcentaje
df["stroke"].value_counts(normalize=True) * 100

stroke
0    95.127202
1     4.872798
Name: proportion, dtype: float64